python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -236,11 +237,26 @@ def _zip_one_block(
         # Swap blocks if ordering was inverted during block alignment splitting.
         block, other_block = other_block, block
     # Zip block and other blocks.
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _try_zip(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
 
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _try_zip(block: Block, other_block: Block) -> Block:
+    try:
+        return BlockAccessor.for_block(block).zip(other_block)
+    except ValueError as e:
+        pyarrow_table = _lazy_import_pyarrow_table()
+        if pyarrow_table:
```

Comment:
- to generalize this conversion from any block type to to Arrow block, can we use `Block.to_arrow()` method?
- we can also avoid the recursive `_try_zip` call and simplify the branching logic by converting the block, then calling `block1.zip(block2)` directly

thinking something along the lines of:
```
def _convert_to_arrow_and_zip(block: Block, other_block: Block) -> Block:
	pyarrow_table = _lazy_import_pyarrow_table()
    if pyarrow_table:
    	if not isinstance(block, pyarrow_table):
    		block = block.to_arrow()
		if not isinstance(other_block, pyarrow_table):
			other_block = other_block.to_arrow()
	try:
        return BlockAccessor.for_block(block).zip(other_block)
    except ValueError as e:
        # Re-raise the ValueError
        raise e
```

Reply:
This doesn't work b/c the underlying type of the `block`s is `pandas.DataFrame`:
```
E                       ray.exceptions.RayTaskError(AttributeError): ray::_zip_one_block() (pid=83910, ip=127.0.0.1)
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/execution/operators/zip_operator.py", line 240, in _zip_one_block
E                           result = _convert_to_arrow_and_zip(block, other_block)
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/execution/operators/zip_operator.py", line 254, in _convert_to_arrow_and_zip
E                           block = block.to_arrow()
E                         File "/Users/michaelhly/miniforge3/envs/ray/lib/python3.9/site-packages/pandas/core/generic.py", line 6204, in __getattr__
E                           return object.__getattribute__(self, name)
E                       AttributeError: 'DataFrame' object has no attribute 'to_arrow'

```

Reply:
I can get rid of the recursive calls

Reply:
@scottjlee how about:
```python3
def _try_zip(block: Block, other_block: Block) -> Block:
    block_acc = BlockAccessor.for_block(block)
    try:
        return block_acc.zip(other_block)
    except ValueError as e:
        pyarrow_table = _lazy_import_pyarrow_table()
        if pyarrow_table:
            if not isinstance(block, pyarrow_table):
                return BlockAccessor.for_block(block_acc.to_arrow()).zip(other_block)
            if not isinstance(other_block, pyarrow_table):
                return block_acc.zip(BlockAccessor.for_block(other_block).to_arrow())

        # Re-raise the ValueError
        raise e
```

Reply:
or:
```python3
def _try_zip(block: BlockAccessor, other_block: Block) -> Block:
    try:
        return block.zip(other_block)
    except ValueError as e:
        pyarrow_table = _lazy_import_pyarrow_table()
        if pyarrow_table:
            if not isinstance(other_block, pyarrow_table):
                other_block = BlockAccessor.for_block(other_block).to_arrow()
                return block.zip(other_block)
            else:
                block = BlockAccessor.for_block(block.to_arrow())
                return _try_zip(block, other_block)

        # Re-raise the ValueError
        raise e
```

Reply:
ah good catch, using the `BlockAccessor.for_block(...).to_arrow()` looks good -- for the second approach, in the final `else` case, I think we can replace the remaining recursive call with `block.zip(other_block)` as well? if that works, we can clean up the logic inside the `if pyarrow_table:` block to use `block.zip(other_block)` once, outside the `try-catch` (then the error will automatically be raised, without the need to catch and re-raise now).

Reply:
@scottjlee I think b/c we're doing the lazy import, we need to re-raise the `ValueError` or it gets swallowed in case `pyarrow` is not installed as a dependency